src/GovUk.Education.ExploreEducationStatistics.Admin/Services/ReleaseDataFileService.cs

```diff
@@ -0,0 +1,532 @@
+        }
+
+        public async Task<Either<ActionResult, Unit>> Delete(Guid releaseId,
+            Guid id,
+            bool forceDelete = false)
+        {
+            return await Delete(releaseId, new List<Guid>
+            {
+                id
+            }, forceDelete: forceDelete);
+        }
+
+        public async Task<Either<ActionResult, Unit>> Delete(Guid releaseId,
+            IEnumerable<Guid> ids,
+            bool forceDelete = false)
+        {
+            return await _persistenceHelper
+                .CheckEntityExists<Release>(releaseId)
+                .OnSuccess(async release => await _userService.CheckCanUpdateRelease(release, ignoreCheck: forceDelete))
+                .OnSuccess(async release =>
+                    await ids.Select(id => _releaseFileRepository.CheckFileExists(releaseId, id, ReleaseFileTypes.Data))
+                        .Aggregate())
+                .OnSuccessVoid(async files =>
+                {
+                    foreach (var file in files)
+                    {
+                        var metaReleaseFileReference =
+                            await GetAssociatedReleaseFileReference(file, Metadata);
+
+                        if (await _releaseFileRepository.FileIsLinkedToOtherReleases(releaseId, file.Id))
+                        {
+                            await _releaseFileRepository.Delete(releaseId, file.Id);
+                            await _releaseFileRepository.Delete(releaseId, metaReleaseFileReference.Id);
+                        }
+                        else
+                        {
+                            await _importService.RemoveImportTableRowIfExists(releaseId, file.Filename);
+                            await _blobStorageService.DeleteBlob(
+                                PrivateFilesContainerName,
+                                file.Path()
+                            );
+                            await _blobStorageService.DeleteBlob(
+                                PrivateFilesContainerName,
+                                metaReleaseFileReference.Path()
+                            );
+
+                            // If this is a replacement then unlink it from the original
+                            if (file.ReplacingId.HasValue)
+                            {
+                                var originalFile = await _fileRepository.Get(file.ReplacingId.Value);
+                                originalFile.ReplacedById = null;
+                                _contentDbContext.Update(originalFile);
+                            }
+
+                            await _releaseFileRepository.Delete(releaseId, file.Id);
+                            await _releaseFileRepository.Delete(releaseId, metaReleaseFileReference.Id);
+
+                            _contentDbContext.ReleaseFileReferences.Remove(file);
+                            _contentDbContext.ReleaseFileReferences.Remove(metaReleaseFileReference);
```

Comment:
Think we can create a `Delete` method out of this stuff in `FileRepository`?

Reply:
I considered that originally but decided against it because we've already fetched the ReleaseFileReferences from the database, and if it's going to be a delete by Id it would have to fetch them again. With it already being in the context it probably doesn't hit the database again so I've made this change now.